pkg/kms/azure/azure.go

```diff
@@ -0,0 +1,240 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azure
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	jose "gopkg.in/square/go-jose.v2"
+
+	kvauth "github.com/Azure/azure-sdk-for-go/services/keyvault/auth"
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/v7.1/keyvault"
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+type KMS struct {
+	client    *keyvault.BaseClient
+	vaultURL  string
+	vaultName string
+	keyName   string
+}
+
+var (
+	errAzureReference = errors.New("kms specification should be in the format azurekms://[VAULT_NAME][VAULT_URL]/[KEY_NAME]")
+
+	referenceRegex = regexp.MustCompile(`^azurekms://([^/]+)/([^/]+)?$`)
+)
+
+const ReferenceScheme = "azurekms://"
+
+func ValidReference(ref string) error {
+	if !referenceRegex.MatchString(ref) {
+		return errAzureReference
+	}
+	return nil
+}
+
+func parseReference(resourceID string) (vaultURL, vaultName, keyName string, err error) {
+	v := referenceRegex.FindStringSubmatch(resourceID)
+	if len(v) != 3 {
+		err = errors.Errorf("invalid azurekms format %q", resourceID)
+		return
+	}
+
+	vaultURL = fmt.Sprintf("https://%s/", v[1])
+	vaultName, keyName = strings.Split(v[1], ".")[0], v[2]
+	return
+}
+
+func NewAzureKMS(ctx context.Context, keyResourceID string) (*KMS, error) {
+	vaultURL, vaultName, keyName, err := parseReference(keyResourceID)
+	if err != nil {
+		return nil, err
+	}
+
+	azureTentatID := os.Getenv("AZURE_TENANT_ID")
+	if azureTentatID == "" {
+		return nil, errors.New("AZURE_TENANT_ID is not set")
+	}
+
+	azureClientID := os.Getenv("AZURE_CLIENT_ID")
```

Comment:
I don't know much about Azure - are these typically set as environment variables? Would it make sense for them to be passed as flags instead?

Reply:
I was following a similar approach that I saw for vault. I will review that when apply the rebate and refactor to the new way

Thanks for you review and feedback

Reply:
Thanks! I asked the same question on the vault one, and I guess that environment variable is the standard, canonical method of specifying the address, so it made sense to keep.

Reply:
I think viper.GetString works with both flags and env vars?

Reply:
@dlorenc @cpanato from what I remember, in Azure the default is by environment variables, alternatively a file could be used or cli device login. https://github.com/Azure/azure-sdk-for-go#more-authentication-details

The env based login approach that is currently implemented is usually used for application based login with either an Azure Application (client secret/cert) or Azure Managed Service Identity (works only on azure based vms and services, where this is available) is used for the login.

I think for the cosign cli it would be valuable to implement also the **auth.NewAuthorizerFromCli()**, which then uses an installed azure cli to execute a device login, where a browser is openend and the user can do his single sign on. With this method not only technical users, but also regular cli users could run the signing. 
Maybe this could be toggled via cli flag.

Reply:
I had that in my mind as well, we also use the env var is in CAPZ (cluster API for azure)

Is that ok to keep as-is for this iteration? @dlorenc

Reply:
I think also that a specific sign in by user might be a use case, that is not that relevant when only a CI is doing the signing. Maybe something to consider for the future :)